Bugsnag/KSCrash/Source/KSCrash/Recording/Tools/BSG_KSMachHeaders.m

```diff
@@ -272,35 +190,145 @@ bool bsg_populate_mach_image_info(const struct mach_header *header, intptr_t sli
     info->uuid = uuid;
     info->name = imageName;
     info->slide = slide;
+    info->removed = FALSE;
+    info->next = NULL;
     
     return true;
 }
 
-/**
- * A callback invoked when dyld loads binary images.  It stores enough relevant info about the
- * image to populate a crash report later.
- *
- * @param header A mach_header structure
- *
- * @param slide A virtual memory slide amount. The virtual memory slide amount specifies the difference between the
- *              address at which the image was linked and the address at which the image is loaded.
- */
-void bsg_mach_binary_image_added(const struct mach_header *header, intptr_t slide)
-{
-    BSG_Mach_Binary_Image_Info info = { 0 };
-    if (bsg_populate_mach_image_info(header, slide, &info)) {
-        bsg_add_mach_binary_image(info);
+void bsg_mach_headers_add_image(const struct mach_header *header, intptr_t slide) {
+    
+    BSG_Mach_Header_Info *newImage = malloc(sizeof(BSG_Mach_Header_Info));
+    if (newImage != NULL && errno != ENOMEM) {
+        if (bsg_mach_headers_populate_info(header, slide, newImage)) {
+            
+            bsg_mach_headers_cache_lock();
+            
+            if (bsg_mach_headers_images_head == NULL) {
+                bsg_mach_headers_images_head = newImage;
+            } else {
+                bsg_mach_headers_images_tail->next = newImage;
+            }
+            bsg_mach_headers_images_tail = newImage;
+            
+            bsg_mach_headers_cache_unlock();
+        }
     }
+    
 }
 
-/**
- * Called when a binary image is unloaded.
- */
-void bsg_mach_binary_image_removed(const struct mach_header *header, intptr_t slide)
-{
-    // Convert header into an info struct
-    BSG_Mach_Binary_Image_Info info = { 0 };
-    if (bsg_populate_mach_image_info(header, slide, &info)) {
-        bsg_remove_mach_binary_image(info.imageVmAddr);
+void bsg_mach_headers_remove_image(const struct mach_header *header, intptr_t slide) {
+    BSG_Mach_Header_Info existingImage = { 0 };
+    if (bsg_mach_headers_populate_info(header, slide, &existingImage)) {
+        for (BSG_Mach_Header_Info *img = bsg_mach_headers_get_images(); img != NULL; img = img->next) {
+            if (img->imageVmAddr == existingImage.imageVmAddr) {
+                img->removed = true;
```

Comment:
Is freeing not a possibility here? I suspect it's not due to the need for thread safety when traversing the linked list but it'd be good to confirm that, and add some docs around why this is the case for future travellers

Reply:
As discussed IRL, by making this non-destructive we're avoiding threading issues. Therefore the memory is only freed if the library is reinstalled. Adding some comments to clarify that.